src/render.rs

```diff
@@ -15,169 +14,217 @@ use matrix_sdk::{
     },
     PossiblyRedactedExt,
 };
+
 use weechat::Weechat;
 
-/// Rendering function for membership events (joins, leaves, bans, profile
-/// changes, etc).
-// TODO: We should not return raw strings here but something like [Block]
-// where Block is (String, [Tags]). Each Block represents one or several lines
-// which have the same tags.
-pub fn render_membership(
-    member: &SyncStateEvent<MemberEventContent>,
-    sender: &WeechatRoomMember,
-    target: &WeechatRoomMember,
-) -> String {
-    use MembershipChange::*;
-    let change_op = member.membership_change();
-
-    let operation = match change_op {
-        None => "did nothing",
-        Error => "caused an error", // must never happen
-        Joined => "has joined the room",
-        Left => "has left the room",
-        Banned => "was banned by",
-        Unbanned => "was unbanned by",
-        Kicked => "was kicked from the room by",
-        Invited => "was invited to the room by",
-        KickedAndBanned => "was kicked and banned by",
-        InvitationRejected => "rejected the invitation",
-        InvitationRevoked => "had the invitation revoked by",
-        ProfileChanged { .. } => "_",
-        NotImplemented => "performed an unimplemented operation",
-        _ => "caused an unhandled membership change",
-    };
+use crate::room_buffer::WeechatRoomMember;
 
-    fn formatted_name(member: &WeechatRoomMember) -> String {
-        match &member.display_name {
-            Some(display_name) => {
-                format!("{name} {color_delim}({color_reset}{user_id}{color_delim}){color_reset}",
-                        name = display_name,
-                        user_id = &member.user_id,
-                        color_delim = Weechat::color("chat_delimiters"),
-                        color_reset = Weechat::color("reset"))
-            }
+/// The rendered version of an event.
+pub struct RenderedEvent {
+    /// The UNIX timestamp of the event.
+    pub message_timestamp: u64,
+    /// The tags of the event,
+    pub tags: Vec<String>,
+    // TODO we need to allow multiple message/tags pairs here.
```

Comment:
Is there a reason (other than extra work) not to make messages/tags a `Vec<(String, Vec<String>)>` immediately? So we don't have to fiddle with it later on.

Reply:
No reason besides the extra work, though there's a bit more figuring out to do before I would change the print methods that take the rendered message to handle multiple message/tag pairs.

We would like to render the content separately but the content obviously doesn't have a timestamp nor sender. So the content needs to be renderable without a prefix (sender/timestamp in the case of normal messages).

How this should be done remains to be seen, two traits, some special casing or two trait methods...